### ISSUE TITLE
Multi zone kubernetes cluster will be recreated if the first worker node is provisioned in a different zone as the cluster

### ISSUE THREAD

**Issue by jfeddern** [IBM-Cloud/terraform-provider-ibm#900]
Hi there,

we are facing an issue while setting up our multi zone cluster. In our case we do not want to use the default worker pool, so we delete it after our cluster and all worker pools are created successfully. 
Actually this comes to a strange behaviour when you provision your cluster in two different zones.

Let's assume the following procedure:

We are creating a cluster in zone fra02. The default worker pool will be created in fra02 and the cluster will be attached to the public and private vlan in fra02. 
Next step is to create new worker pools and attach them to two zones (e.g. fra02 and fra04). Now we are assuming that a worker node which is attached to fra04 is created first.
After the default worker pool is deleted and we are going to re run the cluster creation terraform thinks that the cluster needs to be created again because it seems to be that we are changing the private and public vlan id (please refer to the terraform plan output).

After getting a little bit deeper into this topic we figured out that the issue might be related to the current logic of how the public and private vlan values are retrieved (see the following code block in the file https://github.com/IBM-Cloud/terraform-provider-ibm/blob/7422e2378438b0165c936b1b60b2d3437c64507d/ibm/resource_ibm_container_cluster.go#L593)

```
workerFields, err := wrkAPI.List(clusterID, targetEnv)
	if err != nil {
		return fmt.Errorf("Error retrieving workers for cluster: %s", err)
	}
	if len(workerFields) > 0 {
		d.Set("machine_type", strings.Split(workerFields[0].MachineType, ".encrypted")[0])
		d.Set("public_vlan_id", workerFields[0].PublicVlan)
		d.Set("private_vlan_id", workerFields[0].PrivateVlan)
		if workerFields[0].MachineType != "free" {
			if strings.HasSuffix(workerFields[0].MachineType, ".encrypted") {
				d.Set("disk_encryption", true)
			} else {
				d.Set("disk_encryption", false)
			}
		}
	}
``` 

It seems that the container cluster resource is retrieving the information about the public and private vlan from the first worker node which was created. In our case the first worker node (fra04) is the one which was created in a different zone as the cluster (fra02) was created previously.

There are a couple of solutions to avoid these kind of issues.
1. Instead of retrieving the information about the public and private vlan from the first created worker node, just compare it with the terraform state.
2. Instead of retrieving these information from a worker node I would retrieve them directly from the cluster.
3. Add a parameter to get only worker nodes from the initial created datacenter.

### Terraform Version
Terraform version 0.11.14
Terraform ibm provider version 0.19.0

### Affected Resource(s)
Please list the resources as a list, for example:
- ibm_container_cluster

### Terraform Configuration Files
```hcl
#### INPUT VARIABLES ####

# Global variables
variable "resource_group_id" {}

# Cluster-creation: k8s
variable "kubernetes_cluster_name" {}

variable "kube_version" {}

# VLAN and Subnet variables
variable "public_vlan_id_fra02" {}

variable "private_vlan_id_fra02" {}
variable "public_vlan_id_fra04" {}
variable "private_vlan_id_fra04" {}

variable "private_portable_sn_fra02" {}
variable "public_portable_sn_fra02" {}

variable "private_portable_sn_fra04" {}
variable "public_portable_sn_fra04" {}

## Cluster-Creation: Edge worker pool section
variable "machine_type_worker" {}

variable "machine_type_edge" {}

variable "size_per_zone_worker" {}
variable "size_per_zone_edge" {}

##### GLOBAL CONFIGURATION SHARED BETWEEN CLUSTERS #######
resource "ibm_container_cluster" "k8s" {
  name            = "${var.kubernetes_cluster_name}"
  datacenter      = "fra02"
  machine_type    = "${var.machine_type_edge}"
  hardware        = "dedicated"
  public_vlan_id  = "${var.public_vlan_id_fra02}"
  private_vlan_id = "${var.private_vlan_id_fra02}"

  kube_version = "${var.kube_version}"

  default_pool_size = 1

  subnet_id = ["${var.private_portable_sn_fra02}", "${var.public_portable_sn_fra02}"]
  no_subnet = true

  public_service_endpoint  = "true"
  private_service_endpoint = "true"

  resource_group_id = "${var.resource_group_id}"
}

########### WORKER ###########

resource "ibm_container_worker_pool" "worker" {
  worker_pool_name = "worker"
  machine_type     = "${var.machine_type_worker}"
  cluster          = "${ibm_container_cluster.k8s.id}"
  size_per_zone    = "${var.size_per_zone_worker}"
  hardware         = "dedicated"
  disk_encryption  = "true"

  labels = {
    "dedicated" = "worker"
  }

  resource_group_id = "${var.resource_group_id}"
}

resource "ibm_container_worker_pool_zone_attachment" "worker-fra02" {
  cluster         = "${ibm_container_cluster.k8s.id}"
  worker_pool     = "${element(split("/",ibm_container_worker_pool.worker.id),1)}"
  zone            = "fra02"
  private_vlan_id = "${var.private_vlan_id_fra02}"
  public_vlan_id  = "${var.public_vlan_id_fra02}"

  resource_group_id = "${var.resource_group_id}"
}

resource "ibm_container_worker_pool_zone_attachment" "worker-fra04" {
  cluster         = "${ibm_container_cluster.k8s.id}"
  worker_pool     = "${element(split("/",ibm_container_worker_pool.worker.id),1)}"
  zone            = "fra04"
  private_vlan_id = "${var.private_vlan_id_fra04}"
  public_vlan_id  = "${var.public_vlan_id_fra04}"

  resource_group_id = "${var.resource_group_id}"
}

########### EDGE ###########

resource "ibm_container_worker_pool" "edge" {
  worker_pool_name = "edge"
  machine_type     = "${var.machine_type_edge}"
  cluster          = "${ibm_container_cluster.k8s.id}"
  size_per_zone    = "${var.size_per_zone_edge}"
  hardware         = "dedicated"
  disk_encryption  = "true"

  labels = {
    "dedicated" = "edge"
  }

  resource_group_id = "${var.resource_group_id}"
}

resource "ibm_container_worker_pool_zone_attachment" "edge-fra02" {
  cluster         = "${ibm_container_cluster.k8s.id}"
  worker_pool     = "${element(split("/",ibm_container_worker_pool.edge.id),1)}"
  zone            = "fra02"
  private_vlan_id = "${var.private_vlan_id_fra02}"
  public_vlan_id  = "${var.public_vlan_id_fra02}"

  resource_group_id = "${var.resource_group_id}"
}

resource "ibm_container_worker_pool_zone_attachment" "edge-fra04" {
  cluster         = "${ibm_container_cluster.k8s.id}"
  worker_pool     = "${element(split("/",ibm_container_worker_pool.edge.id),1)}"
  zone            = "fra04"
  private_vlan_id = "${var.private_vlan_id_fra04}"
  public_vlan_id  = "${var.public_vlan_id_fra04}"

  resource_group_id = "${var.resource_group_id}"
}

# By default IBM creates a worker node in a default worker-pool at cluster-creation
# We don't want to use this worker-pool in our setup therefor we have to create the default worker-pool after cluster-creation
resource "null_resource" "delete-default-worker-pool" {
  provisioner "local-exec" {
    command = "ibmcloud ks worker-pool rm --cluster ${ibm_container_cluster.k8s.id} --worker-pool ${ibm_container_cluster.k8s.worker_pools.0.id}"
  }

  depends_on = [
    "ibm_container_worker_pool_zone_attachment.edge-fra02",
    "ibm_container_worker_pool_zone_attachment.edge-fra04",
    "ibm_container_worker_pool_zone_attachment.worker-fra02",
    "ibm_container_worker_pool_zone_attachment.worker-fra04",
  ]
}

# We have to attach the kubernetes cluster to the private portable subnets in zone fra04 after all worker nodes have been provisioned successfully
# Added sleep to give IBM some time to recognize that there are new subnets
resource "null_resource" "attach-private-subnet-to-k8s-cluster" {
  provisioner "local-exec" {
    command = "ibmcloud ks cluster subnet add --cluster ${ibm_container_cluster.k8s.id} --subnet-id ${var.private_portable_sn_fra04} && sleep 300"
  }

  depends_on = [
    "ibm_container_worker_pool_zone_attachment.edge-fra02",
    "ibm_container_worker_pool_zone_attachment.edge-fra04",
    "ibm_container_worker_pool_zone_attachment.worker-fra02",
    "ibm_container_worker_pool_zone_attachment.worker-fra04",
  ]
}

# We have to attach the kubernetes cluster to the public portable subnets in zone fra04 after all worker nodes have been provisioned successfully
# Added sleep to give IBM some time to recognize that there are new subnets
resource "null_resource" "attach-public-subnet-to-k8s-cluster" {
  provisioner "local-exec" {
    command = "ibmcloud ks cluster subnet add --cluster ${ibm_container_cluster.k8s.id} --subnet-id ${var.public_portable_sn_fra04} && sleep 300"
  }

  depends_on = [
    "ibm_container_worker_pool_zone_attachment.edge-fra02",
    "ibm_container_worker_pool_zone_attachment.edge-fra04",
    "ibm_container_worker_pool_zone_attachment.worker-fra02",
    "ibm_container_worker_pool_zone_attachment.worker-fra04",
  ]
}

```

### Expected Behavior
Terraform should not trigger any changes (indeed there were no changes)

### Actual Behavior
Terraform wants to trigger a new cluster creation: 

```hcl
-/+ module.creation.ibm_container_cluster.k8s (new resource required)
       id:                           "mycluster" => <computed> (forces new resource)
       albs.#:                       "4" => <computed>
       billing:                      "hourly" => "hourly"
       datacenter:                   "fra02" => "fra02"
       default_pool_size:            "1" => "1"
       disk_encryption:              "true" => "true"
       hardware:                     "dedicated" => "dedicated"
       ingress_hostname:             "mycluster.eu-de.containers.appdomain.cloud" => <computed>
       ingress_secret:               <sensitive> => <computed> (attribute changed)
       is_trusted:                   "false" => "false"
       isolation:                    "" => <computed>
       kube_version:                 "1.15.5" => "1.15.5"
       machine_type:                 "b3c.4x16" => "b3c.4x16"
       name:                         "mycluster" => "mycluster"
       no_subnet:                    "true" => "true"
       private_vlan_id:              "3456789" => "1234567" (forces new resource)
       public_vlan_id:               "4567890" => "2345678" (forces new resource)
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Create a multi zone cluster
2. Delete the worker node which was provisioned by default during cluster creation
3. Check if the first worker node is attached to a different zone as defined during cluster creation by running `ibmcloud ks worker ls --cluster <clustername>`
4. Run terraform plan again

### Summary
Due to many reason (e.g. changing the machine-type of an existing cluster is not really possible) we do not want to use the default worker pool provided during the cluster creation. So this is no option to provision the cluster with the default worker pool.
Unfortunately the only workaround we could figure out is not to delete the default worker pool, so leave it as it is. Indeed this will increase the costs for our kubernetes cluster as we are using a virtual machine which we are not going to use anyway. 

Kind regards,
Jan

**Comment by hkantare** [comment]
We are working on this

**Comment by hkantare** [comment]
As a workaround you can use the suppress diff on machine_type by adding lifecycle block to container resource
resource "ibm_container_cluster" "test" {
----------
lifecycle {
        ignore_changes = ["machine_type"]
    }
}

**Comment by jfeddern** [comment]
Great thank you! @hkantare

**Comment by Anil-CM** [comment]
we can ensure only the default worker pool's public_vlan_id, private_vlan_id, and disk_encryption are read in the resourceIBMContainerClusterRead() by following change.

`if len(workerPools) > 0 && workerPoolContains(workerPools, defaultWorkerPool) {
		workersByPool, err := wrkAPI.ListByWorkerPool(clusterID, defaultWorkerPool, false, targetEnv)
		if err != nil {
			return fmt.Errorf("Error retrieving workers of default worker pool for cluster: %s", err)
		}

		d.Set("machine_type", strings.Split(workerFields[0].MachineType, ".encrypted")[0])
		d.Set("public_vlan_id", workersByPool[0].PublicVlan)
		d.Set("private_vlan_id", workersByPool[0].PrivateVlan)
		if workerFields[0].MachineType != "free" {
			if strings.HasSuffix(workersByPool[0].MachineType, ".encrypted") {
				d.Set("disk_encryption", true)
			} else {
				d.Set("disk_encryption", false)
			}
		}

		if len(workersByPool) > 0 {
			hardware := workersByPool[0].Isolation
			switch strings.ToLower(hardware) {
			case "":
				hardware = hardwareShared
			case isolationPrivate:
				hardware = hardwareDedicated
			case isolationPublic:
				hardware = hardwareShared
			}
			d.Set("hardware", hardware)
		}

		defaultWorkerPool, err := workerPoolsAPI.GetWorkerPool(clusterID, defaultWorkerPool, targetEnv)
		if err != nil {
			return err
		}
		zones := defaultWorkerPool.Zones
		for _, zone := range zones {
			if zone.ID == cls.DataCenter {
				d.Set("default_pool_size", zone.WorkerCount)
				break
			}
		}
		d.Set("worker_pools", flattenWorkerPools(workerPools))
	}`

**Comment by Anil-CM** [comment]
we can ensure only the default worker pool's public_vlan_id, private_vlan_id, and disk_encryption are read in the resourceIBMContainerClusterRead() by following change.

`workerPools, err := workerPoolsAPI.ListWorkerPools(clusterID, targetEnv)
	if err != nil {
		return err
	}
	if len(workerPools) > 0 && workerPoolContains(workerPools, defaultWorkerPool) {
		workersByPool, err := wrkAPI.ListByWorkerPool(clusterID, defaultWorkerPool, false, targetEnv)
		if err != nil {
			return fmt.Errorf("Error retrieving workers of default worker pool for cluster: %s", err)
		}

		d.Set("machine_type", strings.Split(workersByPool[0].MachineType, ".encrypted")[0])
		d.Set("public_vlan_id", workersByPool[0].PublicVlan)
		d.Set("private_vlan_id", workersByPool[0].PrivateVlan)
		if workersByPool[0].MachineType != "free" {
			if strings.HasSuffix(workersByPool[0].MachineType, ".encrypted") {
				d.Set("disk_encryption", true)
			} else {
				d.Set("disk_encryption", false)
			}
		}`

**Comment by Anil-CM** [comment]
we can ensure only the default worker pool's public_vlan_id, private_vlan_id, and disk_encryption are read in the resourceIBMContainerClusterRead() by following change.

```
workerPools, err := workerPoolsAPI.ListWorkerPools(clusterID, targetEnv)
	if err != nil {
		return err
	}
	if len(workerPools) > 0 && workerPoolContains(workerPools, defaultWorkerPool) {
		workersByPool, err := wrkAPI.ListByWorkerPool(clusterID, defaultWorkerPool, false, targetEnv)
		if err != nil {
			return fmt.Errorf("Error retrieving workers of default worker pool for cluster: %s", err)
		}

		d.Set("machine_type", strings.Split(workersByPool[0].MachineType, ".encrypted")[0])
		d.Set("public_vlan_id", workersByPool[0].PublicVlan)
		d.Set("private_vlan_id", workersByPool[0].PrivateVlan)
		if workersByPool[0].MachineType != "free" {
			if strings.HasSuffix(workersByPool[0].MachineType, ".encrypted") {
				d.Set("disk_encryption", true)
			} else {
				d.Set("disk_encryption", false)
			}
		}
                .
                .
                .
        }
```

**Comment by jfeddern** [comment]
As far as I understood this could work. Just for clarification, if we don't have a defaultWorkerPool deployed, terraform will compare the values of public_vlan_id and private_vlan_id with the current terraform state, right?

**Comment by jfeddern** [comment]
Issue solved with bugfix containing in version 0.20.0.
Thanks for your support!
